src/librustc_mir/hair/cx/mod.rs

```diff
@@ -169,38 +156,46 @@ impl<'a, 'gcx, 'tcx> Cx<'a, 'gcx, 'tcx> {
         };
 
         let clamp = |n| {
-            let size = self.integer_bit_width(ty);
-            trace!("clamp {} with size {} and amt {}", n, size, 128 - size);
-            let amt = 128 - size;
+            let gcx = self.tcx.global_tcx();
+            let param_ty = self.param_env.and(gcx.lift(&ty).unwrap());
+            let bit_width = gcx.layout_of(param_ty).unwrap().size.bits();
+            trace!("clamp {} with size {} and amt {}", n, bit_width, 128 - bit_width);
+            let amt = 128 - bit_width;
             let result = (n << amt) >> amt;
             trace!("clamp result: {}", result);
-            result
+            ConstValue::Scalar(Scalar::Bits {
+                bits: result,
+                defined: bit_width as u8,
+            })
         };
 
         use rustc::mir::interpret::*;
         let lit = match *lit {
             LitKind::Str(ref s, _) => {
                 let s = s.as_str();
                 let id = self.tcx.allocate_bytes(s.as_bytes());
-                let ptr = MemoryPointer::new(id, Size::from_bytes(0));
-                ConstValue::ByValPair(
-                    PrimVal::Ptr(ptr),
-                    PrimVal::from_u128(s.len() as u128),
+                ConstValue::ScalarPair(
+                    Scalar::Ptr(id.into()),
+                    Scalar::Bits {
+                        bits: s.len() as u128,
+                        defined: self.tcx.data_layout.pointer_size.bits() as u8,
+                    }
                 )
             },
             LitKind::ByteStr(ref data) => {
                 let id = self.tcx.allocate_bytes(data);
-                let ptr = MemoryPointer::new(id, Size::from_bytes(0));
-                ConstValue::ByVal(PrimVal::Ptr(ptr))
+                ConstValue::Scalar(Scalar::Ptr(id.into()))
             },
-            LitKind::Byte(n) => ConstValue::ByVal(PrimVal::Bytes(n as u128)),
+            LitKind::Byte(n) => ConstValue::Scalar(Scalar::Bits {
+                bits: n as u128,
+                defined: 8,
+            }),
             LitKind::Int(n, _) if neg => {
                 let n = n as i128;
                 let n = n.overflowing_neg().0;
-                let n = clamp(n as u128);
-                ConstValue::ByVal(PrimVal::Bytes(n))
+                clamp(n as u128)
```

Comment:
Is this a true clamping operation (i.e. saturating), or a truncation? I actually expect truncation.